### PR TITLE
Contribute StandardScaler and MinMax transformations

### PR DESCRIPTION
**Pull request by jmrozanec**
The following PR contribute a StandardScaler transformation, so that we can perform such a transformation on incoming streams considering a given window frame.
---

Checklist

- [x] Code complies with PEP-8 and is consistent with the framework.
- [x] Code is properly documented.
- [x] Tests are included for new functionality or updated accordingly.
- [x] Travis CI build passes with no errors.
- [x] Test Coverage is maintained (threshold is -0.2%).
- [x] Files changed (update, add, delete) are in the PR's scope (no extra files are included).

### CODE REVIEW

File: src/skmultiflow/transform/minmax_scaler.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+        self.__configure()
+
+    def __configure(self):
+        self.window = FastBuffer(max_size=self.window_size)
+
+    def transform(self, X):
+        """ transform
+
+        Does the transformation process in the samples in X.
```
Comment (jacobmontiel) [inline src/skmultiflow/transform/minmax_scaler.py]: This line should be the first in the docstring.

File: src/skmultiflow/transform/minmax_scaler.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            The sample or set of samples that should be transformed.
+
+        """
+        r, c = get_dimensions(X)
+        for i in range(r):
+            row = np.copy([X[i][:]])
+            for j in range(c):
+                value = X[i][j]
+                min = self._get_min(j)
+                max = self._get_max(j)
```
Comment (jacobmontiel) [inline src/skmultiflow/transform/minmax_scaler.py]: `max` is a reserved name in Python. Consider `max_val` and `min_val` for the corresponding variables.

File: src/skmultiflow/transform/minmax_scaler.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+        """
+        X = self.transform(X)
+
+        return X
+
+    def partial_fit(self, X, y=None):
+        """ partial_fit
+
+        Partial fits the model.
```
Comment (jacobmontiel) [inline src/skmultiflow/transform/minmax_scaler.py]: This line should be the first in the docstring.

File: src/skmultiflow/transform/minmax_scaler.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        """ partial_fit_transform
+
+        Partially fits the model and then apply the transform to the data.
+
+        Parameters
+        ----------
+        X: numpy.ndarray of shape (n_samples, n_features)
+            The sample or set of samples that should be transformed.
+
+        y: Array-like
```
Comment (jacobmontiel) [inline src/skmultiflow/transform/minmax_scaler.py]: Please indicate that this parameter is optional.

File: src/skmultiflow/transform/minmax_scaler.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        """ partial_fit
+
+        Partial fits the model.
+
+        Parameters
+        ----------
+        X: numpy.ndarray of shape (n_samples, n_features)
+            The sample or set of samples that should be transformed.
+
+        y: Array-like
```
Comment (jacobmontiel) [inline src/skmultiflow/transform/minmax_scaler.py]: Please indicate that this parameter is optional.

File: src/skmultiflow/transform/minmax_scaler.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+        Partially fits the model and then apply the transform to the data.
+
+        Parameters
+        ----------
+        X: numpy.ndarray of shape (n_samples, n_features)
+            The sample or set of samples that should be transformed.
+
+        y: Array-like
+            The true labels.
```
Comment (jacobmontiel) [inline src/skmultiflow/transform/minmax_scaler.py]: Rephrase suggestion: "The target values." since this can also be used on regression data sets.

File: src/skmultiflow/transform/minmax_scaler.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+        Partial fits the model.
+
+        Parameters
+        ----------
+        X: numpy.ndarray of shape (n_samples, n_features)
+            The sample or set of samples that should be transformed.
+
+        y: Array-like
+            The true labels.
```
Comment (jacobmontiel) [inline src/skmultiflow/transform/minmax_scaler.py]: Rephrase suggestion: "The target values."

File: src/skmultiflow/transform/missing_values_cleaner.py [hunk 8]
```diff
@@ -98,6 +98,8 @@ def transform(self, X):
         
         """
         r, c = get_dimensions(X)
+        print("dimensions: {}x{}".format(r, c)) # TODO delete
```
Comment (jacobmontiel) [inline src/skmultiflow/transform/missing_values_cleaner.py]: This change shouldn't be part of the PR. Applies to both lines.

File: src/skmultiflow/transform/standard_scaler.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
+import numpy as np
+
+from skmultiflow.transform.base_transform import StreamTransform
+from skmultiflow.utils import FastBuffer, get_dimensions
+
+
+class StandardScaler(StreamTransform):
```
Comment (jacobmontiel) [inline src/skmultiflow/transform/standard_scaler.py]: The method name should make explicit that this method is based on data windows. This is to avoid false expectations in the context of incremental learning.

File: src/skmultiflow/transform/standard_scaler.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+        self.__configure()
+
+    def __configure(self):
+        self.window = FastBuffer(max_size=self.window_size)
+
+    def transform(self, X):
+        """ transform
+
+        Does the transformation process in the samples in X.
```
Comment (jacobmontiel) [inline src/skmultiflow/transform/standard_scaler.py]: This line should be the first in the docstring.

File: src/skmultiflow/transform/standard_scaler.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        if not self.window.is_empty():
+            std = np.nanstd(np.array(self.window.get_queue())[:, column_index])
+        if(std == 0.):
+            std = 1.
+        return std
+
+    def partial_fit_transform(self, X, y=None):
+        """ partial_fit_transform
+
+        Partially fits the model and then apply the transform to the data.
```
Comment (jacobmontiel) [inline src/skmultiflow/transform/standard_scaler.py]: This line should be the first in the docstring.

File: src/skmultiflow/transform/standard_scaler.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+        """
+        X = self.transform(X)
+
+        return X
+
+    def partial_fit(self, X, y=None):
+        """ partial_fit
+
+        Partial fits the model.
```
Comment (jacobmontiel) [inline src/skmultiflow/transform/standard_scaler.py]: This line should be the first in the docstring.

File: src/skmultiflow/transform/minmax_scaler.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+    def _get_max(self, column_index):
+        max = 1.
+        if not self.window.is_empty():
+            max = np.nanmax(np.array(self.window.get_queue())[:, column_index])
+        return max
+
+    def partial_fit_transform(self, X, y=None):
+        """ partial_fit_transform
+
+        Partially fits the model and then apply the transform to the data.
```
Comment (jacobmontiel) [inline src/skmultiflow/transform/minmax_scaler.py]: This line should be the first in the docstring.

File: src/skmultiflow/transform/minmax_scaler.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+        y: Array-like
+            The true labels.
+
+        Returns
+        -------
+        MinmaxScaler
+            self
+
+        """
+        X = np.asarray(X)
```
Comment (jacobmontiel) [inline src/skmultiflow/transform/minmax_scaler.py]: According to the docstring, the data type of `X` is a `numpy.ndarray`, why cast it again?

File: src/skmultiflow/transform/standard_scaler.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+        y: Array-like
+            The true labels.
+
+        Returns
+        -------
+        StandardScaler
+            self
+
+        """
+        X = np.asarray(X)
```
Comment (jacobmontiel) [inline src/skmultiflow/transform/standard_scaler.py]: According to the docstring, the data type of `X` is a `numpy.ndarray`, why cast it again?

File: src/skmultiflow/transform/standard_scaler.py [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+                stdev = self._get_std(j)
+                transformed = (value - mean) / stdev
+                X[i][j] = transformed
+            self.window.add_element(row)
+        return X
+
+    def _get_mean(self, column_index):
+        mean = 0.
+        if not self.window.is_empty():
+            mean = np.nanmedian(np.array(self.window.get_queue())[:, column_index])
```
Comment (jacobmontiel) [inline src/skmultiflow/transform/standard_scaler.py]: Should this line use `np.nanmean`?

**Comment by jmrozanec** [comment]
@jacobmontiel thank you for the feedback. Will fix the code based on comments and submit a new PR soon. Best!

**Comment by jmrozanec** [comment]
@jacobmontiel all requested changes were implemented. Thank you for the review!

File: .gitignore [hunk 17]
```diff
@@ -30,6 +30,7 @@ parts/
 sdist/
 var/
 wheels/
+venv/
```
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply .gitignore]: Rolled back.

File: src/skmultiflow/transform/minmax_scaler.py (hunk 4 above)
Comment (jacobmontiel) [reply src/skmultiflow/transform/minmax_scaler.py]: This is what I meant:

```
        y: numpy.ndarray (optional, default=None)
            The target values.
```

File: src/skmultiflow/transform/minmax_scaler.py (hunk 5 above)
Comment (jacobmontiel) [reply src/skmultiflow/transform/minmax_scaler.py]: This is what I meant:

```
        y: numpy.ndarray (optional, default=None)
            The target values.
```

**Comment by jmrozanec** [comment]
@jacobmontiel all required changes were implemented.

**Comment by jacobmontiel** [comment]
Thanks for the changes and patience @jmrozanec

The PR is ready to merge.

**Comment by jmrozanec** [comment]
@jacobmontiel you are welcome! 😄 Glad the features made it to master!
